File: biometric/src/main/java/dev/skomlach/biometric/compat/engine/internal/face/vivo/VivoFaceUnlockModule.java
package dev.skomlach.biometric.compat.engine.internal.face.vivo;

import android.annotation.SuppressLint;
import android.os.Build;

import androidx.annotation.RestrictTo;
import androidx.core.os.CancellationSignal;
import com.vivo.framework.facedetect.FaceDetectManager;
import java.util.Collections;
import dev.skomlach.biometric.compat.engine.AuthenticationFailureReason;
import dev.skomlach.biometric.compat.engine.BiometricInitListener;
import dev.skomlach.biometric.compat.engine.BiometricMethod;
import dev.skomlach.biometric.compat.engine.internal.AbstractBiometricModule;
import dev.skomlach.biometric.compat.engine.internal.core.interfaces.AuthenticationListener;
import dev.skomlach.biometric.compat.engine.internal.core.interfaces.RestartPredicate;
import dev.skomlach.biometric.compat.utils.logging.BiometricLoggerImpl;
import dev.skomlach.common.misc.ExecutorHelper;
import me.weishu.reflection.Reflection;

@RestrictTo(RestrictTo.Scope.LIBRARY)
public class VivoFaceUnlockModule extends AbstractBiometricModule {
    private FaceDetectManager manager = null;

    @SuppressLint("WrongConstant")
    public VivoFaceUnlockModule(BiometricInitListener listener) {
        super(BiometricMethod.FACE_VIVO);
        Reflection.unseal(getContext(), Collections.singletonList("com.vivo.framework.facedetect"));
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            try {
                manager = getContext().getSystemService(FaceDetectManager.class);
            } catch (Throwable ignore) {
                manager = null;
            }
        } else {
            try {
                manager = (FaceDetectManager) getContext().getSystemService("face_detect_service");
            } catch (Throwable ignore) {
                manager = null;
            }
        }
        if (manager == null)
            try {
                manager = FaceDetectManager.getInstance();
            } catch (Throwable ignore) {
                manager = null;
            }
        if (listener != null) {
            listener
                    .initFinished(getBiometricMethod(), VivoFaceUnlockModule.this);
        }
    }

    @Override
    public boolean isManagerAccessible() {
        return manager != null;
    }

    @Override
    public boolean isHardwarePresent() {
        if (manager != null) {
            try {
                return manager.isFaceUnlockEnable();
            } catch (Throwable e) {
                BiometricLoggerImpl.e(e, getName());
            }
        }

        return false;
    }

    @Override
    public boolean hasEnrolled() {
        if (manager != null) {
            try {
                return manager.isFaceUnlockEnable() && manager.hasFaceID();
            } catch (Throwable e) {
                BiometricLoggerImpl.e(e, getName());
            }
        }

        return false;
    }

    @Override
    public void authenticate(final CancellationSignal cancellationSignal,
                             final AuthenticationListener listener,
                             final RestartPredicate restartPredicate) throws SecurityException {

        BiometricLoggerImpl.d(getName() + ".authenticate - " + getBiometricMethod().toString());

        if (manager != null) {
            try {

                final FaceDetectManager.FaceAuthenticationCallback callback =
                        new AuthCallback(restartPredicate, cancellationSignal, listener);

                // Why getCancellationSignalObject returns an Object is unexplained
                final android.os.CancellationSignal signalObject = cancellationSignal == null ? null :
                        (android.os.CancellationSignal) cancellationSignal.getCancellationSignalObject();

                if (signalObject == null)
                    throw new IllegalArgumentException("CancellationSignal cann't be null");
                if (ExecutorHelper.INSTANCE.getExecutor() == null)
                    throw new IllegalArgumentException("Executor cann't be null");
                if (ExecutorHelper.INSTANCE.getHandler() == null)
                    throw new IllegalArgumentException("Handler cann't be null");

                // Occasionally, an NPE will bubble up out of SemBioSomeManager.authenticate
                manager.startFaceUnlock(callback);
                return;
            } catch (Throwable e) {
                BiometricLoggerImpl.e(e, getName() + ": authenticate failed unexpectedly");
            }
        }

        if (listener != null) {
            listener.onFailure(AuthenticationFailureReason.UNKNOWN, tag());
        }
        return;
    }

    class AuthCallback extends FaceDetectManager.FaceAuthenticationCallback {

        private final RestartPredicate restartPredicate;
        private final CancellationSignal cancellationSignal;
        private final AuthenticationListener listener;

        public AuthCallback(RestartPredicate restartPredicate,
                            CancellationSignal cancellationSignal, AuthenticationListener listener) {
            this.restartPredicate = restartPredicate;
            this.cancellationSignal = cancellationSignal;
            this.listener = listener;
        }

        @Override
        public void onFaceAuthenticationResult(int errorCode, int retry_times) {
            BiometricLoggerImpl.d(getName() + ".onFaceAuthenticationResult: " + errorCode + "-" + retry_times);

            if (errorCode == FaceDetectManager.FACE_DETECT_SUCEESS) {
                if (listener != null) {
                    listener.onSuccess(tag());
                }
                return;
            }
            AuthenticationFailureReason failureReason = AuthenticationFailureReason.UNKNOWN;
            switch (errorCode) {
                case FaceDetectManager.FACE_DETECT_NO_FACE:
                    failureReason = AuthenticationFailureReason.SENSOR_FAILED;
                    break;
                case FaceDetectManager.FACE_DETECT_BUSY:
                    failureReason = AuthenticationFailureReason.HARDWARE_UNAVAILABLE;
                    break;
                case FaceDetectManager.FACE_DETECT_FAILED:
                    failureReason = AuthenticationFailureReason.AUTHENTICATION_FAILED;
                    break;
            }

            if (restartPredicate.invoke(failureReason)) {
                if (listener != null) {
                    listener.onFailure(failureReason, tag());
                }
                authenticate(cancellationSignal, listener, restartPredicate);
            } else {
                switch (failureReason) {
                    case SENSOR_FAILED:
                    case AUTHENTICATION_FAILED:
                        lockout();
                        failureReason = AuthenticationFailureReason.LOCKED_OUT;
                        break;
                }

                if (listener != null) {
                    listener.onFailure(failureReason, tag());
                }
            }
        }
    }
}

File: biometric/src/main/java/dev/skomlach/biometric/compat/engine/internal/face/vivo/VivoFaceUnlockModule.kt
/*
 *  Copyright (c) 2021 Sergey Komlach aka Salat-Cx65; Original project: https://github.com/Salat-Cx65/AdvancedBiometricPromptCompat
 *  All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

package dev.skomlach.biometric.compat.engine.internal.face.vivo

import android.annotation.SuppressLint
import androidx.core.os.CancellationSignal
import com.vivo.framework.facedetect.FaceDetectManager
import com.vivo.framework.facedetect.FaceDetectManager.FaceAuthenticationCallback
import dev.skomlach.biometric.compat.AuthenticationFailureReason
import dev.skomlach.biometric.compat.BiometricCryptoObject
import dev.skomlach.biometric.compat.engine.BiometricInitListener
import dev.skomlach.biometric.compat.engine.BiometricMethod
import dev.skomlach.biometric.compat.engine.core.interfaces.AuthenticationListener
import dev.skomlach.biometric.compat.engine.core.interfaces.RestartPredicate
import dev.skomlach.biometric.compat.engine.internal.AbstractBiometricModule
import dev.skomlach.biometric.compat.utils.logging.BiometricLoggerImpl.d
import dev.skomlach.biometric.compat.utils.logging.BiometricLoggerImpl.e


class VivoFaceUnlockModule @SuppressLint("WrongConstant") constructor(listener: BiometricInitListener?) :
    AbstractBiometricModule(BiometricMethod.FACE_VIVO) {
    private var manager: FaceDetectManager? = null

    init {
        manager = try {
            FaceDetectManager.getInstance()
        } catch (e: Throwable) {
            if (DEBUG_MANAGERS)
                e(e, name)
            null
        }


        listener?.initFinished(biometricMethod, this@VivoFaceUnlockModule)
    }

    override val isUserAuthCanByUsedWithCrypto: Boolean
        get() = false

    override fun getManagers(): Set<Any> {
        val managers = HashSet<Any>()
        manager?.let {
            managers.add(it)
        }
        return managers
    }

    override val isManagerAccessible: Boolean
        get() = manager != null
    override val isHardwarePresent: Boolean
        get() {

            try {
                return manager?.isFaceUnlockEnable == true
            } catch (e: Throwable) {
                e(e, name)
            }

            return false
        }

    override fun hasEnrolled(): Boolean {

        try {
            return manager?.isFaceUnlockEnable == true && manager?.hasFaceID() == true
        } catch (e: Throwable) {
            e(e, name)
        }

        return false
    }

    @Throws(SecurityException::class)
    override fun authenticate(
        biometricCryptoObject: BiometricCryptoObject?,
        cancellationSignal: CancellationSignal?,
        listener: AuthenticationListener?,
        restartPredicate: RestartPredicate?
    ) {
        d("$name.authenticate - $biometricMethod; Crypto=$biometricCryptoObject")
        manager?.let {
            try {
                val callback: FaceAuthenticationCallback =
                    AuthCallback(
                        biometricCryptoObject,
                        restartPredicate,
                        cancellationSignal,
                        listener
                    )

                // Why getCancellationSignalObject returns an Object is unexplained
                val signalObject =
                    (if (cancellationSignal == null) null else cancellationSignal.cancellationSignalObject as android.os.CancellationSignal?)
                        ?: throw IllegalArgumentException("CancellationSignal cann't be null")

                // Occasionally, an NPE will bubble up out of SemBioSomeManager.authenticate
                it.startFaceUnlock(callback)
                return
            } catch (e: Throwable) {
                e(e, "$name: authenticate failed unexpectedly")
            }
        }
        listener?.onFailure(AuthenticationFailureReason.UNKNOWN, tag())
        return
    }

    internal inner class AuthCallback(
        private val biometricCryptoObject: BiometricCryptoObject?,
        private val restartPredicate: RestartPredicate?,
        private val cancellationSignal: CancellationSignal?,
        private val listener: AuthenticationListener?
    ) : FaceAuthenticationCallback() {
        override fun onFaceAuthenticationResult(errorCode: Int, retry_times: Int) {
            d("$name.onFaceAuthenticationResult: $errorCode-$retry_times")
            if (errorCode == FaceDetectManager.FACE_DETECT_SUCEESS) {
                listener?.onSuccess(
                    tag(),
                    if(!isUserAuthCanByUsedWithCrypto && isBiometricEnrollChanged)
                        null
                    else
                    BiometricCryptoObject(
                        biometricCryptoObject?.signature,
                        biometricCryptoObject?.cipher,
                        biometricCryptoObject?.mac
                    )
                )
                return
            }
            var failureReason = AuthenticationFailureReason.UNKNOWN
            when (errorCode) {
                FaceDetectManager.FACE_DETECT_NO_FACE -> failureReason =
                    AuthenticationFailureReason.SENSOR_FAILED
                FaceDetectManager.FACE_DETECT_BUSY -> failureReason =
                    AuthenticationFailureReason.HARDWARE_UNAVAILABLE
                FaceDetectManager.FACE_DETECT_FAILED -> failureReason =
                    AuthenticationFailureReason.AUTHENTICATION_FAILED
            }
            if (restartCauseTimeout(failureReason)) {
                authenticate(biometricCryptoObject, cancellationSignal, listener, restartPredicate)
            } else
                if (restartPredicate?.invoke(failureReason) == true) {
                    listener?.onFailure(failureReason, tag())
                    authenticate(
                        biometricCryptoObject,
                        cancellationSignal,
                        listener,
                        restartPredicate
                    )
                } else {
                    if (mutableListOf(
                            AuthenticationFailureReason.SENSOR_FAILED,
                            AuthenticationFailureReason.AUTHENTICATION_FAILED
                        ).contains(failureReason)
                    ) {
                        lockout()
                        failureReason = AuthenticationFailureReason.LOCKED_OUT
                    }
                    listener?.onFailure(failureReason, tag())
                }
        }
    }

}